### PR TITLE
Let OSDK print source line stack traces after panicking

### PR DESCRIPTION
**Pull request by junyang-zh**
In #836 the CI fails but cannot be reproduced locally. Since the PC address of each compilation varies, the current panic stack trace is pretty much not helpful in this case.

In this PR the OSDK will check the QEMU output and automatically print the source line stack trace for the user if the kernel panics.

Here's an example output.

```
[kernel] Hello world from kernel!


   _   ___ _____ ___ ___ ___ _  _   _   ___
  /_\ / __|_   _| __| _ \_ _| \| | /_\ / __|
 / _ \\__ \ | | | _||   /| || .` |/ _ \\__ \
/_/ \_\___/ |_| |___|_|_\___|_|\_/_/ \_\___/


[ERROR]: Uncaught panic!
panicked at /root/asterinas/kernel/aster-nix/src/process/process/job_control.rs:41:5:
a panic for testing
printing stack trace:
   1: fn 0xffffffff886f3dd0 - pc 0xffffffff886f3de8 / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c04b0;

   2: fn 0xffffffff886f38d0 - pc 0xffffffff886f3d87 / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c04c0;

   3: fn 0xffffffff88048030 - pc 0xffffffff8804803a / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c0710;

   4: fn 0xffffffff888606f0 - pc 0xffffffff8886073f / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c0720;

   5: fn 0xffffffff880f12a0 - pc 0xffffffff880f12d5 / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c0770;

   6: fn 0xffffffff880f12e0 - pc 0xffffffff880f133d / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c07c0;

   7: fn 0xffffffff884db6b0 - pc 0xffffffff884db6ed / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c0840;

   8: fn 0xffffffff8851af00 - pc 0xffffffff8851afd8 / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c0870;

   9: fn 0xffffffff881f4a00 - pc 0xffffffff881f4b25 / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c09a0;

  10: fn 0xffffffff884c80e0 - pc 0xffffffff884c832b / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c0b10;

  11: fn 0xffffffff8804a4b0 - pc 0xffffffff8804a975 / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c0c90;

  12: fn 0xffffffff88472750 - pc 0xffffffff8847275e / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c0f40;

  13: fn 0xffffffff88701ab0 - pc 0xffffffff88701ac6 / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c0f60;

  14: fn 0xffffffff882c9b00 - pc 0xffffffff882c9b0e / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c0f80;

  15: fn 0xffffffff88701ab0 - pc 0xffffffff88701ac6 / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c0fb0;

  16: fn 0xffffffff886fdfb0 - pc 0xffffffff886fe014 / registers:

     rax                0x1; rdx 0xffffffff888e5058; rcx               0x16; rbx                0x0;
     rsi 0xffffffff888e4fc0; rdi 0xffff8000004c0678; rbp                0x0; rsp 0xffff8000004c0fd0;

[OSDK] The kernel seems panicked. Parsing source line stack trace:
(  1) /root/asterinas/framework/aster-frame/src/panicking.rs:90
(  2) /root/asterinas/framework/aster-frame/src/panicking.rs:45
(  3) 23xxnbp8ar4351kn:?
(  4) /root/.rustup/toolchains/nightly-2024-01-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/cor4
(  5) 2objm4fch4xr17bm:?
(  6) /root/asterinas/kernel/aster-nix/src/process/process/job_control.rs:51
(  7) /root/asterinas/kernel/aster-nix/src/device/tty/mod.rs:220
(  8) /root/asterinas/kernel/aster-nix/src/process/process/session.rs:106
(  9) /root/asterinas/kernel/aster-nix/src/device/tty/mod.rs:219
( 10) /root/asterinas/kernel/aster-nix/src/process/process/mod.rs:138
( 11) /root/asterinas/kernel/aster-nix/src/lib.rs:110
( 12) /root/.rustup/toolchains/nightly-2024-01-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/cor9
( 13) /root/.rustup/toolchains/nightly-2024-01-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/all0
( 14) /root/asterinas/kernel/aster-nix/src/thread/kernel_thread.rs:38
( 15) /root/.rustup/toolchains/nightly-2024-01-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/all0
( 16) /root/asterinas/framework/aster-frame/src/task/task.rs:277
```

### CODE REVIEW

**Comment by tatetian** [comment]
Looks fine with me. Although I don't see the connection between the `strip_elf` and stack backtrace features.

@StevenJiang1110 Do you have any feedbacks?

**Comment by junyang-zh** [comment]
> Although I don't see the connection between the `strip_elf` and stack backtrace features. Separating them into two atomic commits would be good.

Well if you strip you cannot use `addr2line` to see line numbers. Separating is good, though.

File: osdk/src/cli.rs [hunk 1]
```diff
@@ -264,6 +264,12 @@ pub struct CommonArgs {
         global = true
     )]
     pub linux_x86_legacy_boot: bool,
+    #[arg(
+        long = "strip-elf",
+        help = "Strip the built kernel ELF file for a smaller size",
+        global = true
```
Comment (tatetian) [inline osdk/src/cli.rs:270]: What is the default value in `OSDK.toml`? This is the default value for the CLI, not the configuration file, right?

File: OSDK.toml [hunk 2]
```diff
@@ -27,6 +27,7 @@ args = "$(./tools/qemu_args.sh test)"
 
 [scheme."microvm"]
 boot.method = "qemu-direct"
+build.strip_elf = true
```
Comment (tatetian) [inline OSDK.toml:30]: Why microvm needs to set this value, while other configurations do not?

**Comment by tatetian** [comment]
> > Although I don't see the connection between the `strip_elf` and stack backtrace features. Separating them into two atomic commits would be good.
> 
> Well if you strip you cannot use `addr2line` to see line numbers. Separating is good, though.

Thanks for the explanation.

File: osdk/src/cli.rs (hunk 1 above)
Comment (junyang-zh) [reply osdk/src/cli.rs:270]: For `OSDK.toml` it defaults to false. In CLI the default is false (not set). If set this is true.

File: OSDK.toml (hunk 2 above)
Comment (junyang-zh) [reply OSDK.toml:30]: Because if we don't strip it the ELF would be too large. That's an requirement for microvm only (not for `qemu-direct`).

**Review by StevenJiang1110 — COMMENTED** [review]
This PR looks good to me! I only have some minor comments

File: osdk/src/commands/build/bin.rs [hunk 3]
```diff
@@ -68,38 +68,43 @@ pub fn make_install_bzimage(
     )
 }
 
-pub fn strip_elf_for_qemu(install_dir: impl AsRef<Path>, elf: &AsterBin) -> AsterBin {
-    let stripped_elf_path = {
+pub fn make_elf_for_qemu(install_dir: impl AsRef<Path>, elf: &AsterBin, strip: bool) -> AsterBin {
+    let result_elf_path = {
```
[the remark this answers is not in the document]
Comment (junyang-zh) [reply osdk/src/commands/build/bin.rs:72]: Can't agree. An imperative is better than a suggestion here, as we won't strip if you don't need stripping. It is clear, albeit short.

File: OSDK.toml (hunk 2 above)
Comment (tatetian) [reply OSDK.toml:30]: Ok. Thanks for the explanation.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**